app/knights/knights_basic.py

```diff
@@ -0,0 +1,39 @@
+class Knight:
+    def __init__(self, knight: dict) -> None:
+        self.name = knight["name"]
+        self.power = knight["power"]
+        self.hp = knight["hp"]
+        self.protection = 0
+        self.armour = knight["armour"]
+        self.weapon = knight["weapon"]
+        self.potion = knight["potion"]
+        self.prepare_for_battle()
+
+    def prepare_for_battle(self) -> None:
+        for arm in self.armour:
+            self.protection += arm["protection"]
+
+        # apply weapon
+        self.power += self.weapon["power"]
+
+        # apply potion if exist
+        if self.potion is not None:
+            for key, value in self.potion["effect"].items():

```

Comment:
Use forloop with getattr and setattr https://www.toppr.com/guides/python-guide/questions/what-is-getattr-and-setattr-in-python/